optax/_src/transform.py

```diff
@@ -364,6 +366,44 @@ def update_fn(updates, state, params=None):
   return GradientTransformation(init_fn, update_fn)
 
 
+def scale_by_belief(
+    b1: float = 0.9, b2: float = 0.999,
+    eps: float = 1e-8, eps_root: float = 0.0) -> GradientTransformation:
+  """Rescale updates according to the Adam algorithm.
+
+  References:
+    [Kingma et al, 2014](https://arxiv.org/abs/1412.6980)
+
+  Args:
+    b1: decay rate for the exponentially weighted average of grads.
+    b2: decay rate for the exponentially weighted average of squared grads.
+    eps: term added to the denominator to improve numerical stability.
+    eps_root: term added to the denominator inside the square-root to improve
+      numerical stability when backpropagating gradients through the rescaling.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    mu = jax.tree_map(jnp.zeros_like, params)  # First moment
+    s = jax.tree_map(jnp.zeros_like, params)  # Second Central moment
+    return ScaleByAdamState(count=jnp.zeros([], jnp.int32), mu=mu, nu=s)
+
+  def update_fn(updates, state, params=None):
+    del params
+    mu = _update_moment(updates, state.mu, b1, 1)
+    nu = _update_moment(updates - state.mu, state.nu, b2, 2)
```

Comment:
`updates` and `state.mu` might be arbitrary data structure,
therefore you cannot subtract them directly

Please use tree_multimap as follows:

```
prediction_error = jax.tree_multimap(lambda g, m: g-m, updates, mu)
nu = _update_moment(prediction_error, state.nu, b2, 2)
```